### PR TITLE
Store UPE appearance values in transient

### LINKED ISSUE #2161: Cache UPE theming values
The goal of this issue is to cache the UPE theming values that are generated during checkout. This cache will server 2 purposes: 
- To give a slight performance boost during checkout
- To keep a reference to the themes when they can't be generated at load time (thinking about the My Account > Add Payment Method page).

### PR DESCRIPTION
**Pull request by bborman22**
Fixes #2161 

#### Changes proposed in this Pull Request

This PR adds the functionality to store the calculated UPE appearance values. The background is that to get UPE to visually match the store's theme, we need to do some "calculating" to build an object of CSS rules to pass in. While there is some performance benefit to not calculating that every time, the primary purpose of this PR is to make those calculated CSS rules available to other pages (pay for order, add payment method, update subscription payment method, etc) that use UPE also match the store's theme. The one caveat is that the store's checkout page must first be visited to cache these rules before they can be applied to the other pages mentioned.

This PR checks for the rules in a transient and passes what it finds into JS. In the JS it looks for the rules and if they are not set, calculates them and requests to save them on the server for future use.

#### Testing instructions

- Start with a store with UPE enabled and at least two themes installed. I used Storefront and Astra for my testing.
- With theme 1 selected, visit the Add Payment Method page with a logged in user. My Account > Payment Methods > Add Payment Method.
- Confirm that the UPE form loads, but the fields are not styled to match the theme (assuming checkout page has not been visited yet).
- Leave the add payment method screen without taking any action.
- Add an item to the cart and go to the checkout page. Confirm that the UPE fields are styled to match the theme on the checkout page.
- Without checking out, go back to the Add Payment Method page and confirm that the UPE fields are now styled to match the theme.
- Now as admin, switch to theme 2. This will clear the transient containing the appearance rules.
- Visit the Add Payment Method screen again and confirm that UPE is now back in it's original appearance (not matching either theme 1 or theme 2).
- Visit the checkout page and confirm that UPE is styled to match theme 2.
- Revisit the Add Payment Method screen and confirm UPE is now styled to match theme 2.

All of the above steps can also be taken using the Pay for Order page ([instructions](https://github.com/Automattic/woocommerce-payments/pull/2274)) or Update Subscription Pay Method page ([instructions](https://github.com/Automattic/woocommerce-payments/pull/2386)) to confirm the saved styles are being applied correctly.

**Unstyled Add Payment Method**
Unstyled-UPE-AddPaymentMethod

**Visit Checkout page with styled UPE**
Visit-Checkout-Styled-UPE

**Styled Add Payment Method**
Styled-UPE-AddPaymentMethod

-------------------

- [ ] Added changelog entry (or does not apply)
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

### CODE REVIEW

File: client/checkout/api/index.js [hunk 1]
```diff
@@ -383,6 +383,36 @@ export default class WCPayAPI {
...
+	/**
+	 * Saves the calculated UPE appearance values in a transient.
+	 *
+	 * @param {Object} appearance The UPE appearance object with style values
+	 *
+	 * @return {Promise} The final promise for the request to the server.
+	 */
+	saveUPEAppearance( appearance ) {
+		return this.request( getConfig( 'ajaxUrl' ), {
+			// eslint-disable-next-line camelcase
```
Comment (FangedParakeet) [inline client/checkout/api/index.js:396]:
```suggestion
```
Think this lint comment can be removed, no?

File: client/checkout/classic/upe.js [hunk 2]
```diff
@@ -215,9 +215,15 @@ jQuery( function ( $ ) {
 				const { client_secret: clientSecret, id: id } = response;
 				paymentIntentId = id;
 
-				hiddenElementsForUPE.init();
-				const appearance = getAppearance();
-				hiddenElementsForUPE.cleanup();
+				let appearance = getConfig( 'upeAppareance' );
```
Comment (FangedParakeet) [inline client/checkout/classic/upe.js:218]:
```suggestion
				let appearance = getConfig( 'upeAppearance' );
```
Pedantic suggestion here, but you know I just couldn't let it slide.

File: includes/payment-methods/class-upe-payment-gateway.php [hunk 3]
```diff
@@ -491,11 +497,13 @@ public function process_redirect_payment( $order_id, $intent_id, $save_payment_m
...
-		$payment_fields['accountDescriptor']    = $this->get_account_statement_descriptor();
-		$payment_fields['addPaymentReturnURL']  = wc_get_account_endpoint_url( 'payment-methods' );
-		$payment_fields['gatewayId']            = self::GATEWAY_ID;
-		$payment_fields['paymentMethodsConfig'] = $this->get_enabled_payment_method_config();
+		$payment_fields                           = parent::get_payment_fields_js_config();
+		$payment_fields['accountDescriptor']      = $this->get_account_statement_descriptor();
+		$payment_fields['addPaymentReturnURL']    = wc_get_account_endpoint_url( 'payment-methods' );
+		$payment_fields['gatewayId']              = self::GATEWAY_ID;
+		$payment_fields['paymentMethodsConfig']   = $this->get_enabled_payment_method_config();
+		$payment_fields['upeAppareance']          = get_transient( self::UPE_APPEARANCE_TRANSIENT );
```
Comment (FangedParakeet) [inline includes/payment-methods/class-upe-payment-gateway.php:505]:
```suggestion
		$payment_fields['upeAppearance']          = get_transient( self::UPE_APPEARANCE_TRANSIENT );
```
Apologising again about this one, but you know I had to do it.

File: includes/payment-methods/class-upe-payment-gateway.php [hunk 4]
```diff
@@ -39,6 +39,8 @@ class UPE_Payment_Gateway extends WC_Payment_Gateway_WCPay {
 
 	const METHOD_ENABLED_KEY = 'enabled';
 
+	const UPE_APPEARANCE_TRANSIENT = 'upe_appearance';
```
Comment (FangedParakeet) [inline includes/payment-methods/class-upe-payment-gateway.php:42]:
```suggestion
	const UPE_APPEARANCE_TRANSIENT = 'wcpay_upe_appearance';
```
Feel free to ignore this suggestion, but any thoughts on chucking in a `wcpay` or maybe adding in the gateway ID, to make this transient value stand out a little more?

**Review by FangedParakeet — APPROVED** [review]
I've left a couple suggestions, but they're mostly quite silly, so feel free to adopt or disown them at your own discretion. I tested this with Storefront and Twenty-Twenty-One and this worked perfectly and entirely as expected. Nice one, refreshing to see these other pay pages looking a bit more handsome. Just have a little ponder on the suggestions I've made and act as you see fit, but otherwise this looks fine to me and I offer my blessing and longing to see it merged.

File: client/checkout/classic/upe.js (hunk 2 above)
Comment (bborman22) [reply client/checkout/classic/upe.js:218]: Thank you for catching these! Copy and paste strikes again! 😆

**Pull request merged**